fix: labels values are too long for k8s

olevski opened the pull request:
So I think I came across something similar before but I noticed a bug on Friday through sentry ([link here](https://sentry.dev.renku.ch/organizations/swiss-datascience-center/issues/5210/?environment=renkulab&project=4&referrer=alert_email)) that is the result of a really long project name. We add the project name in the user pod labels but the labels in k8s cannot be longer than 63 characters so when someone with a long project name tries to launch a server it fails.

The way I resolved this is that in the labels I cut down the project name and username to a max of 63 characters. These are only used to filter the user pods when deleting the registry pull secrets. Then I also added the full project name and username in the pod annotations - which have no length limit. The annotations are then used to further filter the pods (that have already been filtered by the k8s api with the labels) when deleting old secrets and it ensures that there are no collisions because of the truncation in the labels.

This is deployed on https://tasko.dev.renku.ch/. I also tested with a project name that is 70 characters long and it works.

I did the acceptance tests locally and they passed after I increased the time for which the tests wait after deleting a project (in order to confirm the project was really deleted). The code here does not affect the deletion of a project at all - for some reason it is taking a bit longer than usual for the project to disappear from the UI after it is deleted in GitLab.

pameladelgado reviewed (commented):
In principle this looks like a good work-around. It's important keep the full length for the pod label matching for the secrets cleanup like you did.

What would happen if you have the same user with two projects that have the same 63 first characters as project name? I guess this is a corner case, but would it be able to launch environments properly?

Couldn't we use the project's GitLab ID? that should be unique enough...

olevski commented:
> What would happen if you have the same user with two projects that have the same 63 first characters as project name?

@pameladelgado this is a good point, but it is not a problem. The renku notebooks api reports the list of active user servers based the label selector for the username and heritage only (i.e. `f"heritage=jupyterhub,renku.io/username={safe_username}"`) and the rest of the filtering is done with the annotations. And the annotations do not have truncated values for anything. So the filtering and everything works.

I tested this out on my deployment with these two projects which have the same first 66 characters:
![image](https://user-images.githubusercontent.com/16360283/108207464-116e3d80-7128-11eb-8fff-5fd3e8a3e6e8.png)
![image](https://user-images.githubusercontent.com/16360283/108207522-28149480-7128-11eb-9989-b37e80edfa53.png)

Everything works as expected.

olevski commented:
> Couldn't we use the project's GitLab ID? that should be unique enough...

@pameladelgado yes that is correct I will change it. It is better to avoid the truncation whenever we can.

olevski commented:
ok redeployed and reran the acceptance tests locally after changing the label to the project id - everything works.

pameladelgado commented:
oh sorry I didnt see your earlier response.
Thanks, I agree that it's better to use the unique ID than to truncate the name. 
I noticed that we are still truncating the username, for this there's the GitLab user ID in case you want to also change this.

ableuler commented on jupyterhub/spawners.py:
```diff
@@ -318,10 +318,10 @@ def get_pod_manifest(self):
+            RENKU_ANNOTATION_PREFIX + "username": safe_username[:63],
             RENKU_ANNOTATION_PREFIX + "commit-sha": options.get("commit_sha"),
-            RENKU_ANNOTATION_PREFIX + "projectName": options.get("project"),
-            "hub.jupyter.org/network-access-hub": "true"
```
This line is necessary for the user session pod to be able to communicate with the hub after network policies were enabled by default in the hub.

olevski replied:
I am sorry I will put that back in. I remember seeing the PR that added that line. Not sure why I deleted it here. Good catch.

ableuler commented on jupyterhub/spawners.py:
```diff
@@ -318,10 +318,10 @@ def get_pod_manifest(self):
-            RENKU_ANNOTATION_PREFIX + "username": safe_username,
+            RENKU_ANNOTATION_PREFIX + "username": safe_username[:63],
             RENKU_ANNOTATION_PREFIX + "commit-sha": options.get("commit_sha"),
-            RENKU_ANNOTATION_PREFIX + "projectName": options.get("project"),
```
Should we keep the project name as an annotation? I understand why you switched to the ID for filtering, but having the name in the annotations could still be useful for us when debugging, etc.

olevski replied:
This is the list of labels @ableuler. The project name already exists in the annotations. I did not remove that.

ableuler replied:
ah sorry, forgot for a moment that we duplicate all labels into annotations already anyway.

ableuler commented on cull_secrets/clean_user_registry_secrets.py:
```diff
@@ -33,12 +33,13 @@
     "renku.io/commit-sha",
     "renku.io/username",
-    "renku.io/projectName",
+    "renku.io/projectId",
```
Maybe we can use gitlabProjectId here (and in other corresponding locations in the code) just to be extra clear?

olevski replied:
Sure I will change the name.

ableuler reviewed (changes requested):
Thanks a lot for identifying this one! Having someone actually go after the sentry errors and hunt down the bugs is really great! 🎖️

olevski commented:
Ok I addressed the comments from @ableuler, kept the username in the labels and added a check for the username length when a notebook is launched as we discussed.

I also ran the acceptance tests locally and they passed. This is also currently deployed at https://tasko.dev.renku.ch/

pameladelgado reviewed (approved):
Looks good to me!
